addons/stock/models/stock_move.py

```diff
@@ -400,7 +400,10 @@ def write(self, vals):
                     lambda m: m.state not in ['draft', 'done', 'cancel'] and float_compare(m.reserved_availability, vals.get('product_uom_qty'), precision_rounding=m.product_uom.rounding) == 1
                 )
                 move_to_unreserve._do_unreserve()
-                (self - move_to_unreserve).filtered(lambda m: m.state == 'assigned').write({'state': 'partially_available'})
+                assigned_moves = (self - move_to_unreserve).filtered(lambda m: m.state == 'assigned')
+                move_partial_available = assigned_moves.filtered((lambda m: m.reserved_availability > 0))
+                move_partial_available.write({'state': 'partially_available'})
+                (assigned_moves - move_partial_available).write({'state': 'confirmed'})
```

Comment:
What about the state 'waiting' ? A move can be 'confirmed' or 'waiting' before the assign.

Reply:
Not sure it would make sense to set 'waiting' because of the signification: 
https://github.com/odoo/odoo/blob/b2d69e9d1e8e0e7464e73098001d103250145f19/addons/stock/models/stock_move.py#L98-L99
In this case, the initial demand is 0 and then the user increases it. With fix applied, after the step 6, the user will then have to lock and check availability again 
Therefore, I think 'Waiting Availability' ('confirmed') is more appropriate

Reply:
Yes my bad, a 'waiting' move will be assign only if the previous move is done. Then it is ok to put it into confirm.

Reply:
I would just add after the if:
(self - move_to_unreserve - receipt_moves_to_reassign)._recompute_state()

Reply:
This will break the new test and two other ones, because `_recompute_state` uses `product_uom_qty` but the latter does not yet have its value updated, so the state will be wrong
For instance, with
https://github.com/odoo/odoo/blob/8f33c7e22f760ca1f7f80be08b520f07a0d1c740/addons/stock/tests/test_move.py#L1106-L1125
L1122 will trigger the method `_recompute_state` and the state will be incorrect:
https://github.com/odoo/odoo/blob/8f33c7e22f760ca1f7f80be08b520f07a0d1c740/addons/stock/models/stock_move.py#L1242-L1243
because we have `move.reserved_availability == move.product_uom_qty == 1`, but `product_uom_qty` is going to become 3 so the state should become `partially_available`

Reply:
After the write then?